pkg/controller/common.go

```diff
@@ -1263,3 +1283,150 @@ func containsString(key string, items []string) bool {
+	if apiErrors.IsNotFound(err) {
+		klog.Infof("Aborting replica management... volume (%s) does not exist", volumeName)
+		return nil
+	} else if err != nil {
+		klog.Errorf("failed to get AzVolume (%s): %v", volumeName, err)
+		return err
+	}
+
+	// replica management should not be executed or retried if AzVolume is scheduled for a deletion or not created.
+	if !isCreated(azVolume) || objectDeletionRequested(azVolume) {
+		klog.Errorf("azVolume (%s) is scheduled for deletion or has no underlying volume object", azVolume.Name)
+		return nil
+	}
+
+	azVolumeAttachments, err := getAzVolumeAttachmentsForVolume(ctx, azr.getClient(), volumeName, replicaOnly)
+	if err != nil {
+		klog.Errorf("failed to list AzVolumeAttachment: %v", err)
+		return err
+	}
+
+	desiredReplicaCount, currentReplicaCount := azVolume.Spec.MaxMountReplicaCount, len(azVolumeAttachments)
+	klog.Infof("control number of replicas for volume (%s): desired=%d,\tcurrent:%d", azVolume.Spec.UnderlyingVolume, desiredReplicaCount, currentReplicaCount)
+
+	if desiredReplicaCount > currentReplicaCount {
+		klog.Infof("Need %d more replicas for volume (%s)", desiredReplicaCount-currentReplicaCount, azVolume.Spec.UnderlyingVolume)
+		if azVolume.Status.Detail == nil || azVolume.Status.State == v1alpha1.VolumeDeleting || azVolume.Status.State == v1alpha1.VolumeDeleted || azVolume.Status.Detail.ResponseObject == nil {
+			// underlying volume does not exist, so volume attachment cannot be made
+			return nil
+		}
+		if err = c.createReplicas(ctx, min(defaultMaxReplicaUpdateCount, desiredReplicaCount-currentReplicaCount), azVolume.Name, azVolume.Status.Detail.ResponseObject.VolumeID, azVolume.Spec.Parameters, azr); err != nil {
+			klog.Errorf("failed to create %d replicas for volume (%s): %v", desiredReplicaCount-currentReplicaCount, azVolume.Spec.UnderlyingVolume, err)
+			return err
+		}
+	}
+	return nil
+}
+func (c *SharedState) createReplicas(ctx context.Context, numReplica int, volumeName, volumeID string, volumeContext map[string]string, azr azReconciler) error {
+	// if volume is scheduled for clean up, skip replica creation
+	if _, cleanUpScheduled := c.cleanUpMap.Load(volumeName); cleanUpScheduled {
+		return nil
+	}
+
+	// get pods linked to the volume
+	pods, err := c.getPodsFromVolume(volumeName)
+	if err != nil {
+		return err
+	}
+
+	// acquire per-pod lock to be released upon creation of replica AzVolumeAttachment CRIs
+	for _, pod := range pods {
+		v, _ := c.podLocks.LoadOrStore(pod, &sync.Mutex{})
+		podLock := v.(*sync.Mutex)
+		podLock.Lock()
+		defer podLock.Unlock()
+	}
+
+	nodes, err := c.getNodesForReplica(ctx, volumeName, pods, numReplica, azr)
+	if err != nil {
+		klog.Errorf("failed to get a list of nodes for replica attachment: %v", err)
+		return err
+	}
+
+	numCreated := 0
+	for _, node := range nodes {
+		if err := createReplicaAzVolumeAttachment(ctx, azr, volumeID, node, volumeContext); err != nil {
```

Comment:
This will handle the case where the CRI creation failed. Can there be cases when the CRI was created successfully but the actual attachment failed and are we adding the pending replicas to the queue for that case as well?

Reply:
Currently, we handle failed attachments for replicas in replica controller by deleting the replica CRI and creating a replacement replica. I suppose we could think about integrating this process into the new queueing method.